master/internal/stream/experiments.go

```diff
@@ -0,0 +1,277 @@
+	}
+}
+
+// ExperimentSubscriptionSpec is what a user submits to define an experiment subscription.
+// determined:streamable
+type ExperimentSubscriptionSpec struct {
+	ExperimentIds []int `json:"experiment_ids"`
+	Since         int64 `json:"since"`
+}
+
+func getExperimentMsgsWithWorkspaceID(expMsgs []*ExperimentMsg) *bun.SelectQuery {
+	q := db.Bun().NewSelect().Model(&expMsgs).
+		Column("id").
+		Column("job_id").
+		Column("owner_id").
+		Column("project_id").
+		Column("unmanaged").
+		Column("state").
+		Column("archived").
+		Column("progress").
+		Column("start_time").
+		Column("end_time").
+		Column("notes").
+		Column("projects.workspace_id").
+		Join("JOIN projects ON experiment_msg.project_id = projects.id")
+	return q
+}
+
+// ExperimentCollectStartupMsgs collects ExperimentMsg's that were missed prior to startup.
+func ExperimentCollectStartupMsgs(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec ExperimentSubscriptionSpec,
+) (
+	[]stream.PreparableMessage, error,
+) {
+	var out []stream.PreparableMessage
+
+	if len(spec.ExperimentIds) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     ExperimentsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetExperimentStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	q := db.Bun().NewSelect().
+		TableExpr("experiments e").
+		Column("e.id").
+		Join("JOIN projects p ON e.project_id = p.id").
+		OrderExpr("e.id ASC")
+
+	q = permFilter(q, accessMap, accessScopes)
+
+	// Ignore tmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.ExperimentIds) > 0 {
+		ws.Include("e.id in (?)", bun.In(spec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	var exist []int64
+	err = q.Scan(ctx, &exist)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	// step 3: hydrate appeared IDs into full ExperimentMsgs
+	var expMsgs []*ExperimentMsg
+	if len(appeared) > 0 {
+		query := getExperimentMsgsWithWorkspaceID(expMsgs).
+			Where("experiment_msg.id in (?)", bun.In(appeared))
+		query = permFilter(query, accessMap, accessScopes)
+		err := query.Scan(ctx, &expMsgs)
+		if err != nil && errors.Cause(err) != sql.ErrNoRows {
+			log.Errorf("error: %v\n", err)
+			return nil, err
+		}
+	}
+
+	// step 4: emit deletions and updates to the client
+	out = append(out, stream.DeleteMsg{
```

Comment:
question (non-blocking & out of scope): Does it make sense for us to send empty delete messages on startup?  We are currently expecting this behavior, but why?

Reply:
I think as long as the client expects one or the other, it will work out. I feel like sending an empty deletion message makes more sense because otherwise we won't know if there was no deletion or if the deletion didn't get sent or was lost in transit